Suspected false positive warning when compiling with GCC 11.x

When compiling with GCC 11.x, a warning is generated about an array overread that doesn't seem to be actually valid. This warning appears only when compiling with `-DSB_TEST` and in particular when the extra elements are included in `sb_fe_t` for runtime quasi-reduction verification. In particular, GCC seems convinced that `MULT_POINT(ctx)` has size 48, but this expands to a reference to a `sb_fe_pair_t` which has size 96 on a 64-bit platform when `SB_TEST` is defined.

Normally I'd be very paranoid about a warning like this, but in this case there's no way for the input pointer to point to something with less than 96 bytes in this situation, and there are a decent number of open GCC bugs about this new class of warnings (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=88443).

```
In file included from /mnt/src/sweet-b/src/sb_sw_lib.c:48:
/mnt/src/sweet-b/src/sb_sw_lib.c: In function 'sb_sw_decompress_public_key':
/mnt/src/sweet-b/src/sb_sw_lib.c:1727:25: warning: 'sb_sw_point_decompress' accessing 96 bytes in a region of size 48 [-Wstringop-overflow=]
 1727 |                        !sb_sw_point_decompress(MULT_POINT(ctx),
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1728 |                                                (sb_word_t) sign, ctx, s);
      |                                                ~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/src/sweet-b/src/sb_error.h:53:49: note: in definition of macro 'SB_ERROR_I'
   53 | #define SB_ERROR_IF(err, cond) ((-(sb_error_t) (cond)) & (sb_error_t) \
      |                                                 ^~~~
/mnt/src/sweet-b/src/sb_sw_lib.c:1727:25: note: referencing argument 1 of type sb_fe_pair_t * const'
 1727 |                        !sb_sw_point_decompress(MULT_POINT(ctx),
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1728 |                                                (sb_word_t) sign, ctx, s);
      |                                                ~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/src/sweet-b/src/sb_error.h:53:49: note: in definition of macro 'SB_ERROR_I'
   53 | #define SB_ERROR_IF(err, cond) ((-(sb_error_t) (cond)) & (sb_error_t) \
      |                                                 ^~~~
/mnt/src/sweet-b/src/sb_sw_lib.c:941:1: note: in a call to function 'sb_sw_point_decompress'
  941 | sb_sw_point_decompress(sb_fe_pair_t point[static const 1],
      | ^~~~~~~~~~~~~~~~~~~~~~
```

--- a/src/sb_sw_lib.c
+++ b/src/sb_sw_lib.c
@@ -938,23 +938,22 @@ sb_sw_point_validate(sb_sw_context_t c[static const 1],
 }
 
 static sb_word_t
-sb_sw_point_decompress(sb_fe_pair_t point[static const 1],
+sb_sw_point_decompress(sb_sw_context_t c[static const 1],
                        const sb_word_t sign,
-                       sb_sw_context_t c[static const 1],
                        const sb_sw_curve_t s[static const 1])
 {
     /* First validate the X coordinate of the point. */
     sb_word_t r = 1;
 
     // 5.6.2.3.4 step 2: unreduced points are not valid.
-    r &= sb_fe_lt(&point->x, &s->p->p);
+    r &= sb_fe_lt(&MULT_POINT(c)->x, &s->p->p);
 
     // The input X value may be 0 on some curves (such as NIST P-256).
     // The modular quasi-reduction routine will change this to P.
-    sb_fe_mod_reduce(&point->x, s->p);
+    sb_fe_mod_reduce(&MULT_POINT(c)->x, s->p);
 
     // Compute y^2 = x^3 + ax + b in C_Y1(c)
-    *C_X1(c) = point->x;
+    *C_X1(c) = MULT_POINT(c)->x;
     sb_sw_curve_y2(c, s);
 
     // Compute the candidate square root
@@ -965,7 +964,7 @@ sb_sw_point_decompress(sb_fe_pair_t point[static const 1],
     sb_fe_mod_negate(C_T5(c), C_Y1(c), s->p);
     sb_fe_ctswap(sign_mismatch, C_Y1(c), C_T5(c));
 
-    point->y = *C_Y1(c);
+    MULT_POINT(c)->y = *C_Y1(c);
 
     return r;
 }
@@ -1724,8 +1723,7 @@ sb_error_t sb_sw_decompress_public_key
 
     sb_fe_from_bytes(MULT_POINT_X(ctx), compressed->bytes, e);
     err |= SB_ERROR_IF(PUBLIC_KEY_INVALID,
-                       !sb_sw_point_decompress(MULT_POINT(ctx),
-                                               (sb_word_t) sign, ctx, s));
+                       !sb_sw_point_decompress(ctx, (sb_word_t) sign, s));
     SB_RETURN_ERRORS(err, ctx);
 
     sb_fe_to_bytes(public->bytes, MULT_POINT_X(ctx), e);

--- a/src/sb_sw_lib_tests.c.h
+++ b/src/sb_sw_lib_tests.c.h
@@ -1179,7 +1179,7 @@ static size_t verify_recover_public_key(sb_sw_public_t keys[4],
 
     extract_sig_components(&m, sig, message, s, e);
 
-    if (sb_sw_point_decompress(MULT_POINT(&m), 0, &m, s)) {
+    if (sb_sw_point_decompress(&m, 0, s)) {
 
         pk_recovery(&m, s);
 
@@ -1191,7 +1191,7 @@ static size_t verify_recover_public_key(sb_sw_public_t keys[4],
         // re-extract scalars...
         extract_sig_components(&m, sig, message, s, e);
 
-        sb_sw_point_decompress(MULT_POINT(&m), 1, &m, s);
+        sb_sw_point_decompress(&m, 1, s);
 
         pk_recovery(&m, s);
 
@@ -1212,7 +1212,7 @@ static size_t verify_recover_public_key(sb_sw_public_t keys[4],
         sb_fe_mod_add(MULT_POINT_X(&m), MULT_POINT_X(&m), C_T5(&m), s->p); //
         // x = x + N
 
-        if (sb_sw_point_decompress(MULT_POINT(&m), 0, &m, s)) {
+        if (sb_sw_point_decompress(&m, 0, s)) {
 
             pk_recovery(&m, s);
 
@@ -1231,7 +1231,7 @@ static size_t verify_recover_public_key(sb_sw_public_t keys[4],
                           s->p); //
             // x = x + N
 
-            sb_sw_point_decompress(MULT_POINT(&m), 1, &m, s);
+            sb_sw_point_decompress(&m, 1, s);
 
             pk_recovery(&m, s);
 

